Evaluate symexprs on load path of cache not write

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128997

When caching is enabled, an internal model fails with
```
assert_size_stride(bmm_9, (17, s0, 512), (54784, 512, 1))
AssertionError: expected size 17==17, stride 57344==54784 at dim=0
```
looking at this model, the exact problem is when the cache is hit on the forward graph, the generated code for backward fails since the strides of the outputs of forward, passed to backward as inputs, are not what we expected.

This PR changes the evaluation logic so that we defer evaluation of output stride exprs to load path as opposed to eagerly doing it on save path.

I have not been able to come up with a unit test repro for this problem.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D58796503](https://our.internmc.facebook.com/intern/diff/D58796503)

oulgen committed:
Evaluate symexprs on load path of cache not write

[ghstack-poisoned]

oulgen committed:
Update on "Evaluate symexprs on load path of cache not write"


When caching is enabled, an internal model fails with
```
assert_size_stride(bmm_9, (17, s0, 512), (54784, 512, 1))
AssertionError: expected size 17==17, stride 57344==54784 at dim=0
```
looking at this model, the exact problem is when the cache is hit on the forward graph, the generated code for backward fails since the strides of the outputs of forward, passed to backward as inputs, are not what we expected.

This PR changes the evaluation logic so that we defer evaluation of output stride exprs to load path as opposed to eagerly doing it on save path.

I have not been able to come up with a unit test repro for this problem.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen committed:
Update on "Evaluate symexprs on load path of cache not write"


When caching is enabled, an internal model fails with
```
assert_size_stride(bmm_9, (17, s0, 512), (54784, 512, 1))
AssertionError: expected size 17==17, stride 57344==54784 at dim=0
```
looking at this model, the exact problem is when the cache is hit on the forward graph, the generated code for backward fails since the strides of the outputs of forward, passed to backward as inputs, are not what we expected.

This PR changes the evaluation logic so that we defer evaluation of output stride exprs to load path as opposed to eagerly doing it on save path.

I have not been able to come up with a unit test repro for this problem.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented on torch/_inductor/compile_fx.py:
```diff
@@ -552,7 +566,15 @@ def compile_fx_inner(
         assert len(context.output_strides) == 0
-        context.output_strides.extend(compiled_graph.output_strides)
+        shape_env = _shape_env_from_inputs(example_inputs)
+        V.graph.sizevars = SizeVarAllocator(shape_env)
```
I don't know if we'd want to assign a SizeVarAllocator to V.graph.sizevars or just use it here? I actually don't know the implications either way... just noting that it was unassigned before.

oulgen replied:
Ah i see, i can do that, but wasnt sure if we needed these sizevars again or not

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -1115,7 +1117,7 @@ class CompiledFxGraph:
     constants: Dict[str, torch.Tensor]
     torchbind_constants: Dict[str, torch._C.ScriptObject]
-    output_strides: Optional[List[Optional[Tuple[int, ...]]]]
+    output_strides: Optional[List[Optional[Tuple[sympy.Expr, ...]]]]
```
nit nit: maybe call them output_strides_exprs? (since they're different from the output_strides that ultimately get attached to the context)

oulgen replied:
I did this at first but it required renaming quite a few things, especially in torch/_guards.py and since python is not typed very well, i did not want to miss one. So, I chose to not do a renaming. (and mypy did not even catch the one in guards.py

masnesral replied:
Sounds good then

oulgen committed:
Update on "Evaluate symexprs on load path of cache not write"


When caching is enabled, an internal model fails with
```
assert_size_stride(bmm_9, (17, s0, 512), (54784, 512, 1))
AssertionError: expected size 17==17, stride 57344==54784 at dim=0
```
looking at this model, the exact problem is when the cache is hit on the forward graph, the generated code for backward fails since the strides of the outputs of forward, passed to backward as inputs, are not what we expected.

This PR changes the evaluation logic so that we defer evaluation of output stride exprs to load path as opposed to eagerly doing it on save path.

I have not been able to come up with a unit test repro for this problem.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen committed:
Update on "Evaluate symexprs on load path of cache not write"


When caching is enabled, an internal model fails with
```
assert_size_stride(bmm_9, (17, s0, 512), (54784, 512, 1))
AssertionError: expected size 17==17, stride 57344==54784 at dim=0
```
looking at this model, the exact problem is when the cache is hit on the forward graph, the generated code for backward fails since the strides of the outputs of forward, passed to backward as inputs, are not what we expected.

This PR changes the evaluation logic so that we defer evaluation of output stride exprs to load path as opposed to eagerly doing it on save path.

I have not been able to come up with a unit test repro for this problem.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen committed:
Update on "Evaluate symexprs on load path of cache not write"


When caching is enabled, an internal model fails with
```
assert_size_stride(bmm_9, (17, s0, 512), (54784, 512, 1))
AssertionError: expected size 17==17, stride 57344==54784 at dim=0
```
looking at this model, the exact problem is when the cache is hit on the forward graph, the generated code for backward fails since the strides of the outputs of forward, passed to backward as inputs, are not what we expected.

This PR changes the evaluation logic so that we defer evaluation of output stride exprs to load path as opposed to eagerly doing it on save path.

I have not been able to come up with a unit test repro for this problem.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -827,11 +841,8 @@ def fx_codegen_and_compile(
-                            )
-                        )
+                        # Convert to string for eval on the load path
+                        output_strides.append(tuple(str(s) for s in out.layout.stride))
```
It would be much better to use the "real" printer that produce_guards is using, as it ensures that we are consistently printing for Python evaluation. That printer is ShapeGuardPrinter. Note that I do think that at present these coincide relatively closely except for relations (but relations ought not to show up in stride computation), but I could easily imagine refactors where they diverge (in particular, if we decide to clean up SYMPY_INTERP by modifying the printer to go directly to what needs to be bound...)

ezyang replied:
If you are not going to use the correct printer or refactor ShapeGuardPrinter so it is usable in this situation, then there is a silent coupling here and you should try your best to make it obvious that this needs to be updated too if someone does change how the produce guards printing works and modifies SYMPY_INTERP. As it stands, this code is very hard to discover and only would get exercised in obscure places.

oulgen commented:
@ezyang Looking at the failing tests, do you know how this change could result in
> AssertionError: Please convert all Tensors to FakeTensors first or instantiate FakeTensorMode with 'allow_non_fake_inputs'. Found in aten.as_strided.default(tensor([...], size=(16,), dtype=torch.uint8), [16], [1])

I'm still investigating too

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -1115,7 +1115,7 @@ class CompiledFxGraph:
     constants: Dict[str, torch.Tensor]
     torchbind_constants: Dict[str, torch._C.ScriptObject]
-    output_strides: Optional[List[Optional[Tuple[int, ...]]]]
+    output_strides: Optional[List[Optional[Tuple[str, ...]]]]
```
Consider documenting the meaning of the string here, either with a type alias or comment

ezyang reviewed (approved):
seems fine ig

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -552,7 +552,21 @@ def compile_fx_inner(
     if context is not None and context.output_strides is not None:
         assert len(context.output_strides) == 0
-        context.output_strides.extend(compiled_graph.output_strides)
+        shape_env = _shape_env_from_inputs(example_inputs)
```
Re https://github.com/pytorch/pytorch/pull/128997#issuecomment-2179114601

I thought this looked kind of suspicious but could not see anything concretely wrong with it

masnesral replied:
What's the right way to retrieve the shape env? In the codecache, I look for it in the TracingContext: https://github.com/pytorch/pytorch/blob/4e03263224af813fbf5e0e745e84c13268c48dc7/torch/_inductor/codecache.py#L793-L801
But I guess that's looking incomplete since most calls in this file are using `detect_fake_mode()`, which does more stuff. Then there's a single call to `_shape_env_from_inputs()` (now a 2nd); the TODO in that method makes it sound like it's strictly for not breaking some unit tests?

oulgen replied:
Pretty silly, the problem was `torch/_guards.py` typed stride as List, but it needs to be a tuple so that result of stride() can be compared to it. Previously, we would hard cast it to tuple but mypy would not error (another L for the mypy) but now that I make it explicit, the type needed to be updated to be tuple again (correctly).

oulgen committed:
Update on "Evaluate symexprs on load path of cache not write"


When caching is enabled, an internal model fails with
```
assert_size_stride(bmm_9, (17, s0, 512), (54784, 512, 1))
AssertionError: expected size 17==17, stride 57344==54784 at dim=0
```
looking at this model, the exact problem is when the cache is hit on the forward graph, the generated code for backward fails since the strides of the outputs of forward, passed to backward as inputs, are not what we expected.

This PR changes the evaluation logic so that we defer evaluation of output stride exprs to load path as opposed to eagerly doing it on save path.

I have not been able to come up with a unit test repro for this problem.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen committed:
Update on "Evaluate symexprs on load path of cache not write"


When caching is enabled, an internal model fails with
```
assert_size_stride(bmm_9, (17, s0, 512), (54784, 512, 1))
AssertionError: expected size 17==17, stride 57344==54784 at dim=0
```
looking at this model, the exact problem is when the cache is hit on the forward graph, the generated code for backward fails since the strides of the outputs of forward, passed to backward as inputs, are not what we expected.

This PR changes the evaluation logic so that we defer evaluation of output stride exprs to load path as opposed to eagerly doing it on save path.

I have not been able to come up with a unit test repro for this problem.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58796503).

oulgen committed:
Update on "Evaluate symexprs on load path of cache not write"


When caching is enabled, an internal model fails with
```
assert_size_stride(bmm_9, (17, s0, 512), (54784, 512, 1))
AssertionError: expected size 17==17, stride 57344==54784 at dim=0
```
looking at this model, the exact problem is when the cache is hit on the forward graph, the generated code for backward fails since the strides of the outputs of forward, passed to backward as inputs, are not what we expected.

This PR changes the evaluation logic so that we defer evaluation of output stride exprs to load path as opposed to eagerly doing it on save path.

I have not been able to come up with a unit test repro for this problem.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D58796503](https://our.internmc.facebook.com/intern/diff/D58796503)

[ghstack-poisoned]

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58796503).

oulgen commented:
@pytorchbot merge
